packages/embedder/embedder.ts

```diff
@@ -46,206 +22,81 @@ tracer.init({
-    console.log(`embedder: completed ${itemKey} -> ${modelTable}`)
+const parseEmbedderMessage = (message: string): MessageToEmbedder => {
+  const {startAt, endAt, ...input} = JSON.parse(message)
+  return {
+    ...input,
+    startAt: startAt ? new Date(startAt) : undefined,
+    endAt: endAt ? new Date(endAt) : undefined
   }
 }
 
-const tick = async (modelManager: ModelManager) => {
-  console.log(`embedder: tick`)
-  const redisClient = getRedisClient()
-  const redlock = new Redlock([redisClient], {
-    driftFactor: 0.01,
-    retryCount: 10,
-    retryDelay: 250,
-    retryJitter: 50,
-    automaticExtensionThreshold: 500
-  })
+const run = async () => {
+  const SERVER_ID = process.env.SERVER_ID
+  if (!SERVER_ID) throw new Error('env.SERVER_ID is required')
+  const embedderChannel = EmbedderChannelId.join(SERVER_ID)
+  const embedderEnabled = parseEnvBoolean(process.env.AI_EMBEDDER_ENABLED)
+  if (!embedderEnabled) {
+    Logger.log('env.AI_EMBEDDER_ENABLED is false. Embedder will not run.')
+    return
+  }
 
-  await redlock
-    .using(['embedder:lock'], 10000, async (signal: RedlockAbortSignal) => {
-      console.log(`embedder: acquired index queue lock`)
-      // N.B. one of the many benefits of using redlock is the using() interface
-      // will automatically extend the lock if these operations exceed the
-      // original redis timeout time
-      await refreshMetadata()
-      await maybeQueueMetadataItems(modelManager)
+  const redis = new RedisInstance(`embedder_${SERVER_ID}`)
+  const isPrimaryEmbedder = await establishPrimaryEmbedder(redis)
+  const modelManager = getModelManager()
+  if (isPrimaryEmbedder) {
+    // only 1 worker needs to perform these on startup
+    await modelManager.maybeCreateTables()
+    await modelManager.removeOldTriggers()
+    await importHistoricalMetadata()
+    resetStalledJobs()
+  }
 
-      if (signal.aborted) {
-        // Not certain which conditions this would happen, it would
-        // happen after operations took place, so nothing much to do here.
-        console.log('embedder: lock was lost!')
-      }
-    })
-    .catch((err: string) => {
-      // Handle errors (including lock acquisition errors)
-      console.error('embedder: an error occurred ', err)
-    })
-  console.log('embedder: index queue lock released')
+  const onMessage = async (_channel: string, message: string) => {
+    const parsedMessage = parseEmbedderMessage(message)
+    await addEmbeddingsMetadata(parsedMessage)
+  }
 
-  // get the highest priority item and embed it
-  await dequeueAndEmbedUntilEmpty(modelManager)
+  // subscribe to consumer group
+  try {
+    await redis.xgroup('CREATE', 'embedderStream', 'embedderConsumerGroup', '$', 'MKSTREAM')
+  } catch (e) {
+    // stream already exists
+  }
 
-  setTimeout(() => tick(modelManager), POLLING_PERIOD_SEC * 1000)
-}
+  const incomingStream = new RedisStream('embedderStream', 'embedderConsumerGroup', embedderChannel)
```

Comment:
+1 It might be a bit easier to understand if the name would reflect that this only triggers generation of the metadata (which then triggers the embedding processing), so maybe we could name it 'embedMetadataStream'?

Reply:
i like it!